robot/packages/cloud_package_manager.go

```diff
@@ -238,58 +195,110 @@ func (m *cloudManager) Cleanup(ctx context.Context) error {
+		expectedPackageDirectories[pkg.thePackage.LocalDataDirectory()] = true
+	}
+
+	topLevelFiles, err := os.ReadDir(m.packagesDir)
 	if err != nil {
 		return err
 	}
-
-	// keep track of known packages by their hashed name from the package id and version.
-	knownPackages := make(map[string]bool)
-
-	// first remove all symlinks to the packages themself.
-	for _, f := range files {
-		if f.Type()&os.ModeSymlink == os.ModeSymlink {
-			// if managed skip removing package
-			if p, ok := m.managedPackages[PackageName(f.Name())]; ok {
-				knownPackages[p.thePackage.SanitizeName()] = true
-				continue
+	// A packageTypeDir is a directory that contains all of the packages for the specified type. ex: packages/ml_models
+	for _, packageTypeDir := range topLevelFiles {
+		if packageTypeDir.Type()&os.ModeDir != os.ModeDir {
+			continue
+		}
+		packageTypeDirName := filepath.Join(m.packagesDir, packageTypeDir.Name())
+		// read all of the packages in the directory and delete those that aren't in expectedPackageDirectories
+		packageDirs, err := os.ReadDir(packageTypeDirName)
+		if err != nil {
+			allErrors = multierr.Append(allErrors, err)
+			continue
+		}
+		for _, packageDir := range packageDirs {
+			packageDirName := filepath.Join(packageTypeDirName, packageDir.Name())
+			_, expectedToExist := expectedPackageDirectories[packageDirName]
+			if !expectedToExist {
+				m.logger.Debug("Removing old package", packageDirName)
+				allErrors = multierr.Append(allErrors, os.RemoveAll(packageDirName))
 			}
+		}
+		// re-read the directory, if there is nothing left in it, delete the directory
+		packageDirs, err = os.ReadDir(packageTypeDirName)
+		if err != nil {
+			allErrors = multierr.Append(allErrors, err)
+			continue
+		}
+		if len(packageDirs) == 0 {
+			allErrors = multierr.Append(allErrors, os.RemoveAll(packageTypeDirName))
+		}
+	}
 
-			m.logger.Infof("Cleaning up unused package link %s", f.Name())
+	m.legacyMlModelSymlinkCleanup()
+	return allErrors
+}
 
-			// Remove logical symlink to package
-			if err := os.Remove(path.Join(m.packagesDir, f.Name())); err != nil {
-				allErrors = multierr.Append(allErrors, err)
-			}
-		}
+// symlink packages/module_name to packages/ml_models/longdataname for backwards compatablility
+func (m *cloudManager) legacyMLModelSymlinkCreation(p config.PackageConfig) error {
+	if err := linkFile(filepath.Join(m.packagesDir, p.Name), p.LocalDataDirectory()); err != nil {
+		m.logger.Errorf("Failed linking ml_model package %s:%s, %s", p.Package, p.Version, err)
+		return err
+	}
+	return nil
+}
+
+// symlink packages/.data to packages/ml_models for backwards compatability
+func (m *cloudManager) legacyCreateMLModelDotDataDirSymlink(p config.PackageConfig) error {
+	if err := linkFile(filepath.Join(m.packagesDir, ".data"), p.LocalDataParentDirectory()); err != nil {
+		m.logger.Errorf("Failed linking packages/.data to packages/ml_models directory, %s", err)
+		return err
 	}
+	return nil
+}
```

Comment:
We could instead see the impact the backwards incompatibility could produce. How many configs have .data in them? Most should be using the existing placeholder in configs and not point to a directory.

Reply:
outdated -- resolving